src/main/java/htsjdk/samtools/cram/structure/Slice.java

```diff
@@ -181,144 +625,135 @@ void craiIndexInitializationCheck() {
+            case 1:
+                // there is only one reference context, all reads placed are either on the same reference
+                // or are unmapped
+                referenceContext = sliceReferenceContexts.iterator().next();
+                break;
+            default:
+                // placed reads on multiple references and/or a combination of placed and unplaced reads
+                referenceContext = ReferenceContext.MULTIPLE_REFERENCE_CONTEXT;
         }
 
-        if (alignmentStart > 0 && referenceContext.isMappedSingleRef() && ref == null) {
-            throw new IllegalArgumentException ("Mapped slice reference is null.");
+        if (referenceContext.isMappedSingleRef()) {
+            AlignmentContext.validateAlignmentContext(
+                    true, referenceContext,
+                    singleRefAlignmentStart,
+                    singleRefAlignmentEnd - singleRefAlignmentStart + 1);
+            return new AlignmentContext(
+                    referenceContext,
+                    singleRefAlignmentStart,
+                    singleRefAlignmentEnd - singleRefAlignmentStart + 1);
+        } else if (referenceContext.isUnmappedUnplaced()) {
+            return AlignmentContext.UNMAPPED_UNPLACED_CONTEXT;
+        } else {
+            return AlignmentContext.MULTIPLE_REFERENCE_CONTEXT;
         }
+    }
 
-        if (alignmentStart > ref.length) {
-            log.error(String.format("Slice mapped outside of reference: seqID=%s, start=%d, counter=%d.",
-                    referenceContext, alignmentStart, globalRecordCounter));
-            throw new RuntimeException("Slice mapped outside of the reference.");
+    private void validateAlignmentSpanForReference(final byte[] referenceBases) {
+        if (alignmentContext.getReferenceContext().isUnmappedUnplaced()) {
+            return;
         }
 
-        if (alignmentStart - 1 + alignmentSpan > ref.length) {
-            log.warn(String.format("Slice partially mapped outside of reference: seqID=%s, start=%d, span=%d, counter=%d.",
-                    referenceContext, alignmentStart, alignmentSpan, globalRecordCounter));
+        if (referenceBases == null &&
+                alignmentContext.getAlignmentStart() > 0 &&
+                alignmentContext.getReferenceContext().isMappedSingleRef()) {
+            throw new CRAMException ("No reference bases found for mapped slice .");
         }
-    }
 
-    public boolean validateRefMD5(final byte[] ref) {
-        if (referenceContext.isMultiRef()) {
-            throw new SAMException("Cannot verify a slice with multiple references on a single reference.");
+        //TODO: CRAMComplianceTest/c1#bounds triggers this (the reads are mapped beyond reference length),
+        // and CRAMEdgeCasesTest.testNullsAndBeyondRef seems to deliberately test that reads that extend
+        // beyond the reference length should be ok ?
+        if (alignmentContext.getAlignmentStart() > referenceBases.length) {
```

Comment:
can you get an NPE here if you have a null referenceBases but not alignmentContext > 0?